aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -2223,67 +2223,34 @@ Tensor chain_matmul(TensorList matrices) {
 /*
 Calculates the Kronecker product between two Tensors.
 */
-Tensor kron(const Tensor& self, const Tensor& other) {
-  /*
-  We can obtain the kron result using tensordot or einsum. The implementation below uses tensordot.
-  In einsum notation suppose we have `self` with dim 4 and `other` with dim 2
-  the result of below tensordot is in einsum 0123, 45 -> 012345.
-  To obtain the correct kron we need to permute and reshape the array.
-  The permutation rule is the following: going from right to left
-  take axes in turn to form the permutation
-  with our example the correct permutation is 012435 and
-  the kron shape is (shape_self[0], shape_self[1], shape_self[3]*shape_other[0],
-  shape_self[4]*shape_other[1])
-  */
-  std::vector<int64_t> self_sizes = self.sizes().vec();
-  std::vector<int64_t> other_sizes = other.sizes().vec();
-  int64_t self_ndim = self.dim();
-  int64_t other_ndim = other.dim();
-  int64_t min_ndim = std::min(self_ndim, other_ndim);
-  int64_t ndim_diff = std::abs(self_ndim - other_ndim);
-
-  std::vector<int64_t> a_axes(self_ndim);
-  std::vector<int64_t> b_axes(other_ndim);
-  std::iota(a_axes.begin(), a_axes.end(), 0);
-  std::iota(b_axes.begin(), b_axes.end(), 0 + self_ndim);
-
-  bool is_a_larger = self_ndim >= other_ndim;
-  std::vector<int64_t> kron_permutation(self_ndim + other_ndim);
-  for (int64_t i = 0; i < ndim_diff; i++) {
-    kron_permutation[i] = is_a_larger ? a_axes[i] : b_axes[i];
-  }
-  for (int64_t i = 0, j = 0; i < min_ndim; i++, j += 2) {
-    kron_permutation[self_ndim + other_ndim - 1 - j] = b_axes[other_ndim - 1 - i];
-    kron_permutation[self_ndim + other_ndim - 1 - j - 1] = a_axes[self_ndim - 1 - i];
-  }
-
-  std::vector<int64_t> result_shape(std::max(self_ndim, other_ndim));
-  for (int64_t i = 0; i < ndim_diff; i++) {
-    result_shape[i] = is_a_larger ? self_sizes[i] : other_sizes[i];
+Tensor& kron_out(Tensor& result, const Tensor& self, const Tensor& other) {
+  auto maxdim = std::max(self.dim(), other.dim());
+  auto pad_self = maxdim - self.dim();
+  auto pad_other = maxdim - other.dim();
+  DimVector a_reshape(2 * maxdim);
```

Comment:
This will be worse than std::vector for 3+d tensors, because it will have to grow original 5-element vector

Reply:
changed to `c10::SmallVector<int64_t, 10>`